File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/CBuilder/StylesSimpleDemo/StylesSimpleDemoStylesDialog.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "StylesSimpleDemoStylesDialog.h"
#include "StylesSimpleDemoData.h"
#include "StylesSimpleDemoEdit.h"
#include "StylesSimpleDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtons"
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxGraphics"
#pragma link "cxInplaceContainer"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxMRUEdit"
#pragma link "cxStyles"
#pragma link "cxTextEdit"
#pragma link "cxTL"
#pragma link "cxCheckBox"
#pragma link "cxCurrencyEdit"
#pragma link "cxDBLookupComboBox"
#pragma link "cxMaskEdit"
#pragma resource "*.dfm"
TStylesSimpleDemoStylesDialogForm *StylesSimpleDemoStylesDialogForm;
//---------------------------------------------------------------------------
__fastcall TStylesSimpleDemoStylesDialogForm::TStylesSimpleDemoStylesDialogForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::FormCreate(TObject *Sender)
{
  TStrings *ALookupItems =
    ((TcxMRUEditProperties*)tlcStyleNames->Properties)->LookupItems;
  TcxStyleRepository *ARespository = StylesSimpleDemoDataDM->StyleRepository;
  ALookupItems->Clear();
  TcxStyle *AStyle;
  for (int I = 0; I < ARespository->Count; I++){
    AStyle = (TcxStyle*)ARespository->Items[I];
    ALookupItems->AddObject(AStyle->Name, AStyle);
  }

/* remove/add the closing slash on this line to disable/enable the following code*/

  RefreshBinding();

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::tlcStyleNamesPropertiesButtonClick(
  TObject *Sender)
{
  TcxStyle *AStyle = GetSelectedStyle();
  if (AStyle != NULL) 
    ChangeStyle(AStyle);
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::btnRestoreClick(TObject *Sender)
{
/* remove/add the closing slash on this line to disable/enable the following code*/

  if (FRestoreDefaults != NULL) FRestoreDefaults(Sender);
  RefreshBinding();

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::tlcStyleNamesPropertiesEditValueChanged(
  TObject *Sender)
{
  int AItemIndex = ((TcxMRUEditProperties*)tlcStyleNames->Properties)->
    LookupItems->IndexOf(((TcxCustomMRUEdit*)Sender)->EditValue);
  TcxStyle *AStyle = (TcxStyle*)((TcxMRUEditProperties*)tlcStyleNames->Properties)->
    LookupItems->Objects[AItemIndex];
  SetCurrentStyle(AStyle, (TStyles)cxTreeList->FocusedNode->Index);
}
//---------------------------------------------------------------------------

TcxStyle* __fastcall TStylesSimpleDemoStylesDialogForm::GetSelectedStyle()
{
  int AItemIndex = ((TcxMRUEditProperties*)tlcStyleNames->Properties)->LookupItems->
    IndexOf(tlcStyleNames->Values[cxTreeList->FocusedNode]);
  if (AItemIndex != -1)
    return (TcxStyle*)((TcxMRUEditProperties*)tlcStyleNames->Properties)->
      LookupItems->Objects[AItemIndex];
  else
    return NULL;
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::RefreshBinding()
{
  TcxStyle *AStyle;
  for (int I = 0; I < cxTreeList->Root->Count; I++){
    AStyle = GetCurrentStyle((TStyles)I);
    if (AStyle != NULL)
      cxTreeList->Root->Items[I]->Values[1] = AStyle->Name;
    else
      cxTreeList->Root->Items[I]->Values[1] = "";
  }
}
//---------------------------------------------------------------------------

TcxStyle* __fastcall TStylesSimpleDemoStylesDialogForm::GetCurrentStyle(
  TStyles AStyleID)
{
  TcxStyle *Result = NULL;
  TcxTreeListStyles *AStyles = StylesSimpleDemoMainForm->cxDBTreeList->Styles;
  switch (AStyleID){
    case sBackground:
      Result = AStyles->Background; break;
    case sBandbackground:
      Result = AStyles->BandBackground; break;
    case sBandContent:
      Result = AStyles->BandContent; break;
    case sBandHeader:
      Result = AStyles->BandHeader; break;
    case sColumnFooter:
      Result = AStyles->ColumnFooter; break;
    case sColumnHeader:
      Result = AStyles->ColumnHeader; break;
    case sContent:
      Result = AStyles->Content; break;
    case sContentEven:
      Result = AStyles->ContentEven; break;
    case sContentOdd:
      Result = AStyles->ContentOdd; break;
    case sFooter:
      Result = AStyles->Footer; break;
    case sInactive:
      Result = AStyles->Indicator; break;
    case sIncSearch:
      Result = AStyles->IncSearch; break;
    case sIndicator:
      Result = AStyles->Indicator; break;
    case sPreview:
      Result = AStyles->Preview; break;
    case sSelection:
       Result = AStyles->Selection;
  }
  return Result;
}
//---------------------------------------------------------------------------

void __fastcall TStylesSimpleDemoStylesDialogForm::SetCurrentStyle(
  TcxStyle *AStyle, TStyles AStyleID)
{
  TcxTreeListStyles *AStyles = StylesSimpleDemoMainForm->cxDBTreeList->Styles;
  switch (AStyleID){
    case sBackground:
      AStyles->Background = AStyle; break;
    case sBandbackground:
      AStyles->BandBackground = AStyle; break;
    case sBandContent:
      AStyles->BandContent = AStyle; break;
    case sBandHeader:
      AStyles->BandHeader = AStyle; break;
    case sColumnFooter:
      AStyles->ColumnFooter = AStyle; break;
    case sColumnHeader:
      AStyles->ColumnHeader = AStyle; break;
    case sContent:
      AStyles->Content = AStyle; break;
    case sContentEven:
      AStyles->ContentEven = AStyle; break;
    case sContentOdd:
      AStyles->ContentOdd = AStyle; break;
    case sFooter:
      AStyles->Footer = AStyle; break;
    case sInactive:
      AStyles->Inactive = AStyle; break;
    case sIncSearch:
      AStyles->IncSearch = AStyle; break;
    case sIndicator:
      AStyles->Indicator = AStyle; break;
    case sPreview:
      AStyles->Preview = AStyle; break;
    case sSelection:
      AStyles->Selection = AStyle;
  }
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/Delphi/StylesSimpleDemo/StylesSimpleDemoStylesDialog.pas
unit StylesSimpleDemoStylesDialog;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxStyles, cxTL, cxTextEdit, cxInplaceContainer, cxControls,
  cxButtonEdit, cxClasses, StdCtrls, cxMRUEdit, cxMaskEdit,
  cxDBLookupComboBox, cxCurrencyEdit, cxMemo, cxCheckBox,
  cxLookAndFeelPainters, cxButtons, cxGraphics, cxCustomData, Menus;

type
  TStyles = (sBackground, sBandbackground, sBandContent, sBandHeader, sColumnFooter,
  sColumnHeader, sContent, sContentEven,  sContentOdd, sFooter, sInactive, sIncSearch,
  sIndicator, sPreview, sSelection);

  TStylesSimpleDemoStylesDialogForm = class(TForm)
    cxTreeList: TcxTreeList;
    tlcStyle: TcxTreeListColumn;
    tlcStyleNames: TcxTreeListColumn;
    lscrip: TLabel;
    cxStyleRepository1: TcxStyleRepository;
    TreeListStyleSheetDevExpress: TcxTreeListStyleSheet;
    cxStyle1: TcxStyle;
    cxStyle2: TcxStyle;
    cxStyle3: TcxStyle;
    cxStyle4: TcxStyle;
    cxStyle5: TcxStyle;
    cxStyle6: TcxStyle;
    cxStyle7: TcxStyle;
    cxStyle8: TcxStyle;
    cxStyle9: TcxStyle;
    cxStyle10: TcxStyle;
    cxStyle11: TcxStyle;
    cxStyle12: TcxStyle;
    cxStyle13: TcxStyle;
    btnRestore: TcxButton;
    procedure FormCreate(Sender: TObject);
    procedure tlcStyleNamesPropertiesButtonClick(Sender: TObject);
    procedure btnRestoreClick(Sender: TObject);
    procedure tlcStyleNamesPropertiesEditValueChanged(Sender: TObject);
  private
    FRestoreDefaults: TNotifyEvent;
    function GetSelectedStyle: TcxStyle;
    procedure RefreshBinding;
    function GetCurrentStyle(AStyleID: TStyles): TcxStyle;
    procedure SetCurrentStyle(const AStyle: TcxStyle; AStyleID: TStyles);
  public
    property RestoreDefaults: TNotifyEvent read FRestoreDefaults write FRestoreDefaults;
  end;

var
  StylesSimpleDemoStylesDialogForm: TStylesSimpleDemoStylesDialogForm;

implementation

uses StylesSimpleDemoData, StylesSimpleDemoEdit, StylesSimpleDemoMain;

{$R *.dfm}

function ChangeStyleBinding(ACallback: TNotifyEvent): Boolean;
begin
  with TStylesSimpleDemoStylesDialogForm.Create(Application) do
  try
    RestoreDefaults := ACallback;
    Result := ShowModal = mrOK;
  finally
    Free;
  end;
end;

procedure TStylesSimpleDemoStylesDialogForm.FormCreate(Sender: TObject);
var
  i: integer;
  AStyle: TcxStyle;
begin
  with TcxMRUEditProperties(tlcStyleNames.Properties).LookupItems, StylesSimpleDemoDataDM do
  begin
    Clear;
    for i := 0 to StyleRepository.Count -1  do
    begin
       AStyle := TcxStyle(StyleRepository[i]);
       AddObject(AStyle.Name, AStyle);
    end;
  end;
{ remove/add the closing brace on this line to disable/enable the following code}

  RefreshBinding;

//}
end;

procedure TStylesSimpleDemoStylesDialogForm.tlcStyleNamesPropertiesButtonClick(
  Sender: TObject);
var
  AStyle: TcxStyle;
begin
  AStyle := GetSelectedStyle;
  if AStyle <> nil then
    ChangeStyle(AStyle);
end;

procedure TStylesSimpleDemoStylesDialogForm.btnRestoreClick(
  Sender: TObject);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  if Assigned(FRestoreDefaults) then FRestoreDefaults(Sender);
  RefreshBinding;

//}
end;

procedure TStylesSimpleDemoStylesDialogForm.RefreshBinding;
var
  I: Integer;
  AStyle: TcxStyle;
begin
  for i:=0 to cxTreeList.Root.Count - 1 do
  begin
    AStyle := GetCurrentStyle(TStyles(i));
    if AStyle <> nil then
      cxTreeList.Root.Items[i].Values[1] := AStyle.Name
    else
      cxTreeList.Root.Items[i].Values[1] := '';
  end;
end;

function TStylesSimpleDemoStylesDialogForm.GetSelectedStyle: TcxStyle;
var
  ItemIndex: Integer;
begin
  Result := nil;
  ItemIndex := TcxMRUEditProperties(tlcStyleNames.Properties).LookupItems.
    IndexOf(tlcStyleNames.Values[cxTreeList.FocusedNode]);
  if ItemIndex <> -1 then
    Result := TcxStyle(TcxMRUEditProperties(tlcStyleNames.Properties).LookupItems.Objects[ItemIndex]);
end;

function TStylesSimpleDemoStylesDialogForm.GetCurrentStyle(AStyleID: TStyles): TcxStyle;
begin
  Result := nil;
  with StylesSimpleDemoMainForm do
  case AStyleID of
    sBackground:
      Result := cxDBTreeList.Styles.Background;
    sBandbackground:
      Result := cxDBTreeList.Styles.BandBackground;
    sBandContent:
      Result := cxDBTreeList.Styles.BandContent;
    sBandHeader:
      Result := cxDBTreeList.Styles.BandHeader;
    sColumnFooter:
      Result := cxDBTreeList.Styles.ColumnFooter;
    sColumnHeader:
      Result := cxDBTreeList.Styles.ColumnHeader;
    sContent:
      Result := cxDBTreeList.Styles.Content;
    sContentEven:
      Result := cxDBTreeList.Styles.ContentEven;
    sContentOdd:
      Result := cxDBTreeList.Styles.ContentOdd;
    sFooter:
      Result := cxDBTreeList.Styles.Footer;
    sInactive:
      Result := cxDBTreeList.Styles.Indicator;
    sIncSearch:
      Result := cxDBTreeList.Styles.IncSearch;
    sIndicator:
      Result := cxDBTreeList.Styles.Indicator;
    sPreview:
      Result := cxDBTreeList.Styles.Preview;
    sSelection:
       Result := cxDBTreeList.Styles.Selection;
  end;
end;

procedure TStylesSimpleDemoStylesDialogForm.SetCurrentStyle(
  const AStyle: TcxStyle; AStyleID: TStyles);
begin
  with StylesSimpleDemoMainForm do
    case AStyleID of
      sBackground:
        cxDBTreeList.Styles.Background := AStyle;
      sBandbackground:
        cxDBTreeList.Styles.BandBackground := AStyle;
      sBandContent:
        cxDBTreeList.Styles.BandContent := AStyle;
      sBandHeader:
        cxDBTreeList.Styles.BandHeader := AStyle;
      sColumnFooter:
        cxDBTreeList.Styles.ColumnFooter := AStyle;
      sColumnHeader:
        cxDBTreeList.Styles.ColumnHeader := AStyle;
      sContent:
        cxDBTreeList.Styles.Content := AStyle;
      sContentEven:
        cxDBTreeList.Styles.ContentEven := AStyle;
      sContentOdd:
        cxDBTreeList.Styles.ContentOdd := AStyle;
      sFooter:
        cxDBTreeList.Styles.Footer := AStyle;
      sInactive:
        cxDBTreeList.Styles.Inactive := AStyle;
      sIncSearch:
        cxDBTreeList.Styles.IncSearch := AStyle;
      sIndicator:
        cxDBTreeList.Styles.Indicator := AStyle;
      sPreview:
        cxDBTreeList.Styles.Preview := AStyle;
      sSelection:
        cxDBTreeList.Styles.Selection := AStyle;
    end;
end;

procedure TStylesSimpleDemoStylesDialogForm.tlcStyleNamesPropertiesEditValueChanged(
  Sender: TObject);
var
  ItemIndex: Integer;
  AStyle: TcxStyle;
begin
  ItemIndex := TcxMRUEditProperties(tlcStyleNames.Properties).
    LookupItems.IndexOf(TcxCustomMRUEdit(Sender).EditValue);
  AStyle := TcxStyle(TcxMRUEditProperties(tlcStyleNames.Properties).LookupItems.Objects[ItemIndex]);
  SetCurrentStyle(AStyle, TStyles(cxTreeList.FocusedNode.Index));
end;

end.
